### ISSUE TITLE
Unable to run wasm module built from Rust with C++ dependency. Error: Can't resolve 'env'

### ISSUE THREAD

**Issue by JanKaul** [rustwasm/wasm-pack#741]
## 🐛 Bug description
`npm run start` fails when running a wasm module that was built from Rust with C++ dependency using `wasm-pack`. Inspecting the generated wasm module shows that it tries to import `(import "env" "double_it" (func $double_it (type $t5)))` from a module "env" that it doesn't know ("double_it" is the function in the C++ dependency). The C++ dependency is actually a C file which has been renamed to have a ".cpp" ending. `npm run start` works with the ".c" file. So maybe the ".cpp" ending triggers different compilation flags that yield the different result. Since both C and C++ compile to wasm I thought there should be no difference in the usage.

A same problem has been initially encountered also for the ".c" file in the comment: [https://github.com/rustwasm/wasm-pack/issues/621#issuecomment-481929042](https://github.com/rustwasm/wasm-pack/issues/621#issuecomment-481929042). The problem was potentially resolved with a newer clang or Rust update.

Project available here: [https://github.com/JanKaul/simple-wasm](https://github.com/JanKaul/simple-wasm)

#### 🤔 Expected Behavior
`npm run start` should succeed.

#### 👟 Steps to reproduce

1. C++ dependency build with `cc` crate, which determines compiler automatically (I guess `clang`)
1. `wasm-pack build` succeeds
1. `npm install` succeeds
1. `npm run strart` fails with following error:

```
ERROR in ../pkg/simple_wasm_bg.wasm
Module not found: Error: Can't resolve 'env' in '/home/user/workspace/rust/simple-wasm/pkg'
 @ ../pkg/simple_wasm_bg.wasm
 @ ../pkg/simple_wasm.js
 @ ./index.js
 @ ./bootstrap.js
ℹ ｢wdm｣: Failed to compile.
```

#### 🌍 Your environment
`wasm-pack 0.8.1`
`rustc 1.39.0 (4560ea788 2019-11-04)`
```
clang version 9.0.0-2 (tags/RELEASE_900/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by GreatIndieDeveloper** [comment]
You will have to have a C interface (with extern C) so that you can call wasm functions from and you will have to get a custom malloc that works with wasm

**Comment by JanKaul** [comment]
Thank you @GreatIndieDeveloper  for your reply. But as far as I understand both points should be fine.

Your first point is that I need a C interface. But as far as I understood, the C interface is defined in the file `simple-wasm/double-sys/src/lib.rs`. So it is there.

And the second point is that I should provide a malloc function. Since I'm not using any includes from the c or c++ standard library, which normally implements malloc, I thought that I don't have to provide that functionality. The code is just a simple function that works on integers, which doesn't require any OS functions.

Did I get your points right? Or did you mean something else?

And shouldn't it have the same problems with the file having a `.c` ending?

Correct me if I'm wrong.

**Comment by JanKaul** [comment]
Okay the problem seems to be the fact that c++ uses a different ABI and the symbol for the function has a mangled name, that the rust compiler doesn't know. One has to explicitly tell the rust compiler the mangled function name it has to link. This can be done with the attribute `#[link_name = "\u{1}_Z9double_itj"]` before the function definition.
If you use bindgen to generate the bindings, it does it automatically for a c++ file.

**Comment by TjeuKayim** [comment]
When I compile below C code with the `cc` crate, the WASM output contains `(import "env" "malloc" (func $malloc (type 3)))`.
```c
#include <stdlib.h>
int* foo(int a) {
    return malloc(a);
}
```

**Comment by JanKaul** [comment]
It seems like it cannot find the function malloc to link it into the .wasm file. That's why it uses the import "env". Are you compiling your file as a c++ file? If yes you I guess you should use `.cpp_link_stdlib("c++")` to link it against the clang standard library or `.cpp_set_stdlib("stdc++")` to link it against the GNU standard library. I hope it helps.

**Comment by TjeuKayim** [comment]
I was not using c++ but want to compile sqlite. I got a `Module not found: Error: Can't resolve 'env'` just like you. The `wasm32-unknown-unknown` target has no standard c library, and `wasm-pack` does not support wasi or emscripten.

**Comment by cherryleafroad** [comment]
I was trying to compile `tree-sitter` into WASM using the official Rust bindings for it. After a ton of errors during compile, I manually fixed up the build command to make it compile into wasm, although wasm-pack doesn't support target `wasm32-unknown-emscripten`, so I had to manually compile it and hack together a workable example for the js and npm.

Anyways, I did that, and it comes up with this exact error. I used the default example of the `alert()` function too.
